File: <100/051/051.py
import time
import sympy


def replace_digits(n: int, which: int, to: int) -> int:
    nl = list(str(n))
    ws = str(which)
    ts = str(to)

    f = lambda c: ts if c == ws else c
    nl_ = map(f, nl)
    s = "".join(nl_)
    if s[0] == '0':
        return -1
    return int(s)


def generate_replaceds(n: int, which: int) -> list[int]:
    return [replace_digits(n, which, to) for to in range(10)]


def is_ok(n: int, which: int) -> bool:
    l = generate_replaceds(n, which)
    filtered = list(filter(sympy.isprime, l))
    length = len(filtered)
    return length == 8


def main() -> int:
    upper_limit = 10**7
    for n in range(56004, upper_limit):
        for which in range(2):
            if is_ok(n, which):
                return n
    
    return -1


if __name__ == "__main__":
    start = time.time()
    answer = main()
    elapsed = time.time() - start

    print(f'{elapsed} sec')
    print(answer)

File: <100/051/051.jl
using Primes

function to_char(n :: Int) :: Char
  return string(n)[1]
end


function replace_digits(n :: Int, which :: Int, to :: Int) :: Int
  f = collect ∘ string
  vn :: Vector{Char} = f(n)
  wc :: Char = to_char(which)
  tc :: Char = to_char(to)
  g = c -> c == wc ? tc : c

  vn_ = map(g, vn)
  s = string(vn_...)
  if s[1] == '0'
    return -1
  else
    return parse(Int, s)
  end
end


function generate_replaceds(n:: Int, which :: Int) :: Vector{Int}
  return [replace_digits(n, which, to) for to ∈ 0:9]
end


function is_ok(n :: Int, which :: Int) :: Bool
  v = generate_replaceds(n, which)
  filtered = filter(isprime, v)
  len = length(filtered)
  return len == 8
end


function main() :: Nothing
  upper_limit = 10^7
  for n ∈ 56004:upper_limit
    for which ∈ 0:1
      if is_ok(n, which)
        println(n)
        return nothing
      end
    end
  end

  println(-1)
  return nothing
end


@time main()
